### ISSUE TITLE
TENDL photonuclear processing issue for Ra226

### ISSUE THREAD

**Issue by whaeck** [njoy/NJOY2016#201]
An email was posted to the MCNP user forum concerning an issue with a photonuclear ACE file for Ra226 from TENDL 2019.

The length of the ACE file did not correspond to the length written to the xsdir entry, indicating an issue while printing the ACE file. When performing an ACER run for testing an existing ACE file, ACER crashes due to an expected end of file (again due to the ACE file being incomplete).

The following input and ENDF file illustrate the issue:
[input.txt](https://github.com/njoy/NJOY2016/files/6535009/input.txt)
[g-Ra226.tendl.txt](https://github.com/njoy/NJOY2016/files/6535010/g-Ra226.tendl.txt)

**Comment by whaeck** [comment]
I have put this through the feature/pn-iaea branch. That branch contains some diagnostics for checking the locator positions in the photonuclear ACE files when NJOY writes them out (these diagnostics were previously available for continuous energy neutron and charged particle files but this branch will extend it to the photonuclear data). When running this version of NJOY2016, I'm getting an error message on a mismatching locator (i.e. a locator points to a position before the current position in the xss array) which would lead to a malformed ACE file. This new version therefore errors out.

I have now narrowed it down to the MF6 MT51 entry in the Ra226 photonuclear file. In this piece of the ENDF file, there are three reaction products: a neutron, a residual Ra225 and a photon. For some reason, ACER is not counting this reaction as a photon producing reaction when it fills out the IXS array in the particle production blocks. However, when ACER is filling in the MTRH, TYRH, LSIGH, SIGH, etc. blocks for the photon it does pick up the photon from MF6 MT51. Because this offsets the size of the MTRH, TYRH and LSIGH block, what is supposed to be the locator for the cross section of the first MT on the MTRH photon block is in fact the TYR value for a shifted reaction. Since TYR=-1, this results in a locator for that first reaction photon production cross section pointing to a position before the SIGH block.

Long story short: if you go into the evaluation, and set the LAW=0 for ZAP=88225 in MF6 MT51 (second subsection), then the issue goes away in the above mentioned branch. The diagnostics still warn about a locator mismatch later in the file (LANDH and ANDH for photons seem to be correct but the LDLWH block is shifted by ~100 values) which indicates a gap in the file.

**Comment by whaeck** [comment]
The reaction count issue was caused by the LAW=4 entry. Whenever LAW=4 is used, NJOY will rewind the tape to the LAW=2 section but it won't advance the tape back to the subsection it was reading when it rewound the tape. As a result NJOY detected the following subsections in MF6 MT51:
- neutron, subsection 1
- Ra225 with LAW=4, subsection 2, therefore rewind but forget to go back past subsection 2
- Ra225 with LAW=4, subsection 3 (actually subsection 2)

By properly skipping the subsections up to the one that caused the rewind, the reaction counting issue goes away.

The gap before the LDLWH block is however still there. This one seems to be caused by a poor choice for the initialisation value of a temporary variable when generating the angular distributions in the ANDH block for photons. Again, this is due to that LAW=4 subsection in MF6 MT51.

The issue is the izarec variable defined on line ~889 in acepn.f90. It is initialised to 0 but set to a different value when LAW=4, after which (you guessed it) NJOY rewinds the tape back to the first section.

The following code is excuted:
```
                  izarec=0
                  awprec=0
                  if (izap.eq.ip.and.law.eq.4) then
                     izarec=izap
                     awprec=awp
                     ! removed code: rewind, read first subsection, extract izap, awp and law
                  endif

                  !--law2 angular distribution
                  !--also used for law 4 two-body recoils
                  if ((izap.eq.ip.and.law.eq.2).or.&
                    (izarec.eq.ip.and.law.eq.2)) then
                    ! code removed
                  endif
```

When processing the ip particle type for photons (ip=0), NJOY loops over the MF6 subsection for MT51. There it finds the first subsection for izap=1. and LAW=2. Since this is not LAW=4, the first if statement does not get executed. However, in the second if statement, one of the conditions is actually true:
- izap.eq.ip.and.law.eq.2 is false (ip == 0, izap == 1)
- izarec.eq.ip.and.law.eq.2 is TRUE (izarec == 0, ip == 0)

As a result, the neutron angular distribution data is assumed to be for the secondary photon. NJOY processes this subsection and adds data to the ANDH block for secondary photons.

Afterwards, when NJOY processes data to be put in the DLWH block for photons, it finds that the photons in MF6 MT51 use LAW=1, which causes NJOY to reset the LANDPH entry for this MT number to 0 (where it previously had been set to 1 in the previous block due to the neutron data being mistaked for photon data). The ANDH block is however not empty at this point. This causes the gap in the ACE file between the end of the LANDH block and the beginning of the LDLWH block. 

The feature/pn-iaea branch already handles this and the resulting ACE file can therefore be read properly. NJOY2016.63 however does not. Setting izarec to a more appropriate value (like -1) solves the issue.

Due to the urgent nature of these fixes, these have been implemented in a separate branch: fix/tendl-pn

**Comment by jchsublet** [comment]
@whaeck interesting posting and solution for an (g,ng), likely to be short lived though as the original evaluation is a s30 (explicit) version of TENDL-2019, and not a more robust (for acer) s0 (implicit). In that case I wonder how MCNP handle a 10 18; sub-actinide fission channel?  Useful diagnosis though as it impact all g-tendl s30 evaluation.
@kahlerac remember the weird MCNP numerical warning when using the CoNDERC gamma test suite with the s30 g-tendl in contrast to the s0 based one
